packages/driver/src/cy/commands/sessions.ts

```diff
@@ -0,0 +1,822 @@
+
+    return _.uniq(
+      _.flatten(await Promise.map(
+        ([] as string[]).concat(origins), async (v) => {
+          if (v === '*') {
+            return _.keys(await Cypress.backend('get:renderedHTMLOrigins')).concat([current_origin])
+          }
+
+          if (v === 'current_origin') return current_origin
+
+          return $Location.create(v).origin
+        },
+      )),
+    ) as string[]
+  }
+
+  async function _setStorageOnOrigins (originOptions) {
+    const specWindow = cy.state('specWindow')
+
+    const current_origin = $Location.create(window.location.href).origin
+
+    const current_origin_options_index = _.findIndex(originOptions, { origin: current_origin })
+
+    if (current_origin_options_index !== -1) {
+      const opts = originOptions.splice(current_origin_options_index, 1)[0]
+
+      if (!_.isEmpty(opts.localStorage)) {
+        if (opts.localStorage.clear) {
+          window.localStorage.clear()
+        }
+
+        _.each(opts.localStorage.value, (val, key) => localStorage.setItem(key, val))
+      }
+
+      if (opts.sessionStorage) {
+        if (opts.sessionStorage.clear) {
+          window.sessionStorage.clear()
+        }
+
+        _.each(opts.sessionStorage.value, (val, key) => sessionStorage.setItem(key, val))
+      }
+    }
+
+    if (_.isEmpty(originOptions)) {
+      return
+    }
+
+    const origins = originOptions.map((v) => v.origin) as string[]
+
+    const iframes: JQuery<HTMLElement>[] = []
+
+    const $iframeContainer = $(`<div style="display:none"></div>`).appendTo($('body', specWindow.document))
+
+    // if we're on an https domain, there is no way for the secure context to access insecure origins from iframes
+    // since there is no way for the app to access localStorage on insecure contexts, we don't have to clear any localStorage on http domains.
+    if (current_origin.startsWith('https:')) {
+      _.remove(origins, (v) => v.startsWith('http:'))
+    }
+
+    _.each(origins, (u) => {
+      const $iframe = $(`<iframe src="${`${u}/__cypress/automation/setLocalStorage?${u}`}"></iframe>`)
+
+      $iframe.appendTo($iframeContainer)
+      iframes.push($iframe)
+    })
+
+    let onPostMessage
+
+    const successOrigins = [] as string[]
+
+    await new Promise((resolve) => {
+      onPostMessage = (event) => {
+        const data = event.data
+
+        if (data.type === 'set:storage:load') {
+          if (!event.source) {
+            throw new Error('failed to get localStorage')
+          }
+
+          const opts = _.find(originOptions, { origin: event.origin })!
+
+          event.source.postMessage({ type: 'set:storage:data', data: opts }, '*')
+        } else if (data.type === 'set:storage:complete') {
+          successOrigins.push(event.origin)
+          if (successOrigins.length === origins.length) {
+            resolve()
+          }
+        }
+      }
+
+      specWindow.addEventListener('message', onPostMessage)
+    })
+    .timeout(2000)
```

Comment:
might be worth commenting: in what error condition would this timeout ever be reached? the thread for the other origin being locked for over two seconds?

Reply:
85d85eaeeb